### PR TITLE
Add zksync data to safes

### PR DESCRIPTION
**Pull request by alicecha**
# Thank you for contributing to Spellbook!
Please refer to the top of the `readme` in the root of Spellbook to learn how to contribute to Spellbook on DuneSQL.

### CODE REVIEW

**Comment by alicecha** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by Hosuke** [comment]
Hi @alicecha, you can run `git merge main` to update this PR.

**Comment by danielpartida** [comment]
> Hi @alicecha, you can run `git merge main` to update this PR.
 
Hi @Hosuke , I just manually triggered a `git merge main`. Let me know if that was fine or if you need a further action points from our end

File: models/safe/zksync/safe_zksync_schema.yml [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+      blockchain: zksync
+      project: safe
+      contributors: ['danielpartida' , 'kryptaki']
+    freshness:
+      warn_after: { count: 12, period: hour }
+      error_after: { count: 24, period: hour }
+    config:
+      tags: ['safe', 'zksync']
+    description: "Safe addresses"
+    columns:
```
Comment (Hosuke) [inline models/safe/zksync/safe_zksync_schema.yml:15]:
```suggestion
    description: "Safe addresses"
    tests:
      - dbt_utils.unique_combination_of_columns:
          combination_of_columns:
            - block_date
            - address
    columns:
```

File: models/safe/zksync/safe_zksync_schema.yml [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+      blockchain: zksync
+      project: safe
+      contributors: ['danielpartida' , 'kryptaki']
+    freshness:
+      warn_after: { count: 12, period: hour }
+      error_after: { count: 24, period: hour }
+    config:
+      tags: ['safe', 'transfers', 'zksync']
+    description: "Eth transfers for safes"
+    columns:
```
Comment (Hosuke) [inline models/safe/zksync/safe_zksync_schema.yml:46]:
```suggestion
    description: "Eth transfers for safes"
    tests:
      - dbt_utils.unique_combination_of_columns:
          combination_of_columns:
            - block_date
            - tx_hash
            - trace_address
            - amount_raw
    columns:
```

File: models/safe/zksync/safe_zksync_schema.yml [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+      blockchain: zksync
+      project: safe
+      contributors: ['danielpartida' , 'kryptaki']
+    freshness:
+      warn_after: { count: 12, period: hour }
+      error_after: { count: 24, period: hour }
+    config:
+      tags: ['safe', 'zksync']
+    description: "Safe transactions"
+    columns:
```
Comment (Hosuke) [inline models/safe/zksync/safe_zksync_schema.yml:89]:
```suggestion
    description: "Safe transactions"
    tests:
      - dbt_utils.unique_combination_of_columns:
          combination_of_columns:
            - block_date
            - tx_hash
            - trace_address
    columns:
```

**Review by Hosuke — CHANGES_REQUESTED** [review]
We can add some uniqueness tests for incremental models.

File: models/safe/zksync/safe_zksync_schema.yml (hunk 1 above)
Comment (Hosuke) [reply models/safe/zksync/safe_zksync_schema.yml:15]: Since I cannot directly commit to this PR, I may need your help to update the schema.yml.

Comment (alicecha) [reply models/safe/zksync/safe_zksync_schema.yml:15]: Sounds good, I'll push the changes this morning

**Comment by alicecha** [comment]
> We can add some uniqueness tests for incremental models.

@Hosuke I've committed your changes - hopefully this works now?

**Review by Hosuke — APPROVED** [review]
LGTM.✅
Thank you @alicecha 

By the way, the current `safe_zksync.safes` is empty. Is this expected?
https://dune.com/queries/3117217

File: models/safe/zksync/safe_zksync_eth_transfers.sql [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    and et."from" != et.to -- exclude calls to self to guarantee unique key property
+    and et.success = true
+    and (lower(et.call_type) not in ('delegatecall', 'callcode', 'staticcall') or et.call_type is null)
+    and et.value > UINT256 '0' -- et.value is uint256 type
+{% if not is_incremental() %}
+where et.block_time > timestamp '2023-09-01' -- for initial query optimisation
+{% endif %}
+{% if is_incremental() %}
+-- to prevent potential counterfactual safe deployment issues we take a bigger interval
+where et.block_time > date_trunc('day', now() - interval '10' day)
```
Comment (Hosuke) [inline models/safe/zksync/safe_zksync_eth_transfers.sql:37]: Also for incremental filters, we can rewrite in the equivalent way now:
```suggestion
where {{ incremental_predicate('et.block_time') }}
```

**Comment by alicecha** [comment]
> LGTM.✅ Thank you @alicecha
> 
> By the way, the current `safe_zksync.safes` is empty. Is this expected? https://dune.com/queries/3117217

No...I'll take a look later.

**Comment by alicecha** [comment]
> > LGTM.✅ Thank you @alicecha
> > By the way, the current `safe_zksync.safes` is empty. Is this expected? https://dune.com/queries/3117217
> 
> No, this is weird...I'll take a look later.

Quick update @Hosuke 

This where clause is the reason the table is blank: `and et.call_type = 'delegatecall'`

Unlike all other chains zksync does not have call_type delegatecall, and we are seeking advice from their team to make sure we extract the right set of data. I will let you know when we have a clear answer.

**Comment by Hosuke** [comment]
You can run `git merge main` to update the branch.

File: models/safe/zksync/safe_zksync_eth_transfers.sql [hunk 5]
```diff
@@ -0,0 +1,62 @@
+{{
+    config(
+        materialized='incremental',
+        tags = ['dunesql'],
+        alias= alias('eth_transfers'),
```
Comment (Hosuke) [inline models/safe/zksync/safe_zksync_eth_transfers.sql:5]: Same as above.
```suggestion
        schema = 'safe_zksync',
        alias = 'eth_transfers',
```

File: models/safe/zksync/safe_zksync_safes.sql [hunk 6]
```diff
@@ -0,0 +1,48 @@
+{{
+    config(
+        materialized='incremental',
+        tags = ['dunesql'],
+        alias= alias('safes'),
```
Comment (Hosuke) [inline models/safe/zksync/safe_zksync_safes.sql:5]: Same as above.

File: models/safe/zksync/safe_zksync_transactions.sql [hunk 7]
```diff
@@ -0,0 +1,59 @@
+{{
+    config(
+        materialized='incremental',
+        tags = ['dunesql'],
+        alias= alias('transactions'),
```
Comment (Hosuke) [inline models/safe/zksync/safe_zksync_transactions.sql:5]: Same as above.

**Comment by jeff-dude** [comment]
we're unable to merge main into your branch, please do again to see the CI gh action run

**Comment by alicecha** [comment]
> > > LGTM.✅ Thank you @alicecha
> > > By the way, the current `safe_zksync.safes` is empty. Is this expected? https://dune.com/queries/3117217
> > 
> > 
> > No, this is weird...I'll take a look later.
> 
> Quick update @Hosuke
> 
> This where clause is the reason the table is blank: `and et.call_type = 'delegatecall'`
> 
> Unlike all other chains zksync does not have call_type delegatecall, and we are seeking advice from their team to make sure we extract the right set of data. I will let you know when we have a clear answer.

Hi @Hosuke , @jeff-dude, I just pushed some changes which should have solved the empty table issue we had for `safe_zksync.safes`.  We're still working on the `safe_zksync.transactions` table with zksync. The rest should all be working. Thanks for your patience!

File: models/safe/zksync/safe_zksync_safes.sql [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    evt_block_time as creation_time,
+    evt_tx_hash
+from {{ source('gnosis_safe_zksync', 'GnosisSafeProxyFactoryv1_3_0_evt_ProxyCreation') }}
+where 
+    {% if not is_incremental() %}
+    and et.block_time > TIMESTAMP '2023-09-01' -- for initial query optimisation
+    {% endif %}
+    {% if is_incremental() %}
+    and et.block_time > date_trunc('day', now() - interval '7' day)
+    {% endif %}
```
Comment (Hosuke) [inline models/safe/zksync/safe_zksync_safes.sql:37]: We can remove extra `and` here.
```suggestion
where 
    {% if not is_incremental() %}
    et.block_time > TIMESTAMP '2023-09-01' -- for initial query optimisation
    {% endif %}
    {% if is_incremental() %}
    {{ incremental_predicate('et.block_time') }}
    {% endif %}
```

**Comment by alicecha** [comment]
> > > > LGTM.✅ Thank you @alicecha
> > > > By the way, the current `safe_zksync.safes` is empty. Is this expected? https://dune.com/queries/3117217
> > > 
> > > 
> > > No, this is weird...I'll take a look later.
> > 
> > 
> > Quick update @Hosuke
> > This where clause is the reason the table is blank: `and et.call_type = 'delegatecall'`
> > Unlike all other chains zksync does not have call_type delegatecall, and we are seeking advice from their team to make sure we extract the right set of data. I will let you know when we have a clear answer.
> 
> Hi @Hosuke , @jeff-dude, I just pushed some changes which should have solved the empty table issue we had for `safe_zksync.safes`. We're still working on the `safe_zksync.transactions` table with zksync. The rest should all be working. Thanks for your patience!

@Hosuke , @jeff-dude - the code has been updated. We removed the transactions view which is still pending review from this pull request. Let me know if you can merge into the dune spellbook. Thanks!

**Review by jeff-dude — APPROVED** [review]
thank you 🙏

**Pull request merged**